Dialogue client generation fails if doc string contains the `$` symbol

## What happened?

If a doc string contains the `$` symbol then code generation will fail.

```
Exception in thread "main" picocli.CommandLine$ExecutionException: Error while running command (com.palantir.conjure.java.cli.ConjureJavaCli$GenerateCommand@1bb5a082): java.lang.IllegalArgumentException: index 1 for '$i' not in range (received 0 arguments)
        at picocli.CommandLine.executeUserObject(CommandLine.java:1778)
        at picocli.CommandLine.access$900(CommandLine.java:145)
        at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2150)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2144)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2108)
        at picocli.CommandLine$AbstractParseResultHandler.handleParseResult(CommandLine.java:1968)
        at picocli.CommandLine.parseWithHandlers(CommandLine.java:2349)
        at picocli.CommandLine.run(CommandLine.java:2807)
        at picocli.CommandLine.run(CommandLine.java:2744)
        at com.palantir.conjure.java.cli.ConjureJavaCli.main(ConjureJavaCli.java:48)
  Caused by: java.lang.IllegalArgumentException: index 1 for '$i' not in range (received 0 arguments)
        at com.squareup.javapoet.Util.checkArgument(Util.java:53)
        at com.squareup.javapoet.CodeBlock$Builder.add(CodeBlock.java:287)
        at com.squareup.javapoet.TypeSpec$Builder.addJavadoc(TypeSpec.java:439)
        at com.palantir.conjure.java.services.dialogue.DialogueEndpointsGenerator.lambda$endpointField$1(DialogueEndpointsGenerator.java:82)
```

Can be reproduced by adding a `$` symbol to any doc string.

## What did you want to happen?

Client generation should be consistent with other generators.

--- a/changelog/@unreleased/pr-919.v2.yml
+++ b/changelog/@unreleased/pr-919.v2.yml
@@ -1,0 +1,5 @@
+type: fix
+fix:
+  description: Properly escape object documentation strings
+  links:
+  - https://github.com/palantir/conjure-java/pull/919

--- a/conjure-java-core/src/integrationInput/java/com/palantir/product/ManyFieldExample.java
+++ b/conjure-java-core/src/integrationInput/java/com/palantir/product/ManyFieldExample.java
@@ -94,7 +94,7 @@ public final class ManyFieldExample {
     }
 
     /**
-     * docs for items field
+     * docs for items field with exciting character$ used by javapoet.
      */
     @JsonProperty("items")
     public List<String> getItems() {
@@ -287,7 +287,7 @@ public final class ManyFieldExample {
         }
 
         /**
-         * docs for items field
+         * docs for items field with exciting character$ used by javapoet.
          */
         @JsonSetter(value = "items", nulls = Nulls.SKIP, contentNulls = Nulls.FAIL)
         public Builder items(@Nonnull Iterable<String> items) {
@@ -297,7 +297,7 @@ public final class ManyFieldExample {
         }
 
         /**
-         * docs for items field
+         * docs for items field with exciting character$ used by javapoet.
          */
         public Builder addAllItems(@Nonnull Iterable<String> items) {
             ConjureCollections.addAll(this.items, Preconditions.checkNotNull(items, "items cannot be null"));
@@ -305,7 +305,7 @@ public final class ManyFieldExample {
         }
 
         /**
-         * docs for items field
+         * docs for items field with exciting character$ used by javapoet.
          */
         public Builder items(String items) {
             this.items.add(items);

--- a/conjure-java-core/src/integrationInput/java/test/prefix/com/palantir/product/ManyFieldExample.java
+++ b/conjure-java-core/src/integrationInput/java/test/prefix/com/palantir/product/ManyFieldExample.java
@@ -95,7 +95,7 @@ public final class ManyFieldExample {
     }
 
     /**
-     * docs for items field
+     * docs for items field with exciting character$ used by javapoet.
      */
     @JsonProperty("items")
     public List<String> getItems() {
@@ -289,7 +289,7 @@ public final class ManyFieldExample {
         }
 
         /**
-         * docs for items field
+         * docs for items field with exciting character$ used by javapoet.
          */
         @JsonSetter(value = "items", nulls = Nulls.SKIP)
         public Builder items(@Nonnull Iterable<String> items) {
@@ -299,7 +299,7 @@ public final class ManyFieldExample {
         }
 
         /**
-         * docs for items field
+         * docs for items field with exciting character$ used by javapoet.
          */
         public Builder addAllItems(@Nonnull Iterable<String> items) {
             ConjureCollections.addAll(this.items, Preconditions.checkNotNull(items, "items cannot be null"));
@@ -307,7 +307,7 @@ public final class ManyFieldExample {
         }
 
         /**
-         * docs for items field
+         * docs for items field with exciting character$ used by javapoet.
          */
         public Builder items(String items) {
             this.items.add(items);

--- a/conjure-java-core/src/main/java/com/palantir/conjure/java/types/BeanBuilderGenerator.java
+++ b/conjure-java-core/src/main/java/com/palantir/conjure/java/types/BeanBuilderGenerator.java
@@ -278,7 +278,8 @@ public final class BeanBuilderGenerator {
         boolean shouldClearFirst = false;
         return MethodSpec.methodBuilder(prefix + StringUtils.capitalize(field.name))
                 .addJavadoc(Javadoc.render(definition.getDocs(), definition.getDeprecated())
-                        .orElse(""))
+                        .map(rendered -> CodeBlock.of("$L", rendered))
+                        .orElseGet(() -> CodeBlock.builder().build()))
                 .addAnnotations(ConjureAnnotations.deprecation(definition.getDeprecated()))
                 .addModifiers(Modifier.PUBLIC)
                 .returns(builderClass)
@@ -490,7 +491,8 @@ public final class BeanBuilderGenerator {
         FieldDefinition definition = enriched.conjureDef();
         return MethodSpec.methodBuilder(enriched.poetSpec().name)
                 .addJavadoc(Javadoc.render(definition.getDocs(), definition.getDeprecated())
-                        .orElse(""))
+                        .map(rendered -> CodeBlock.of("$L", rendered))
+                        .orElseGet(() -> CodeBlock.builder().build()))
                 .addAnnotations(ConjureAnnotations.deprecation(definition.getDeprecated()))
                 .addModifiers(Modifier.PUBLIC)
                 .returns(builderClass);

--- a/conjure-java-core/src/main/java/com/palantir/conjure/java/types/UnionGenerator.java
+++ b/conjure-java-core/src/main/java/com/palantir/conjure/java/types/UnionGenerator.java
@@ -277,7 +277,8 @@ public final class UnionGenerator {
                             .addJavadoc(Javadoc.render(
                                             entry.getKey().getDocs(),
                                             entry.getKey().getDeprecated())
-                                    .orElse(""))
+                                    .map(rendered -> CodeBlock.of("$L", rendered))
+                                    .orElseGet(() -> CodeBlock.builder().build()))
                             .addAnnotations(ConjureAnnotations.deprecation(
                                     entry.getKey().getDeprecated()))
                             .addModifiers(Modifier.PUBLIC, Modifier.ABSTRACT)

--- a/conjure-java-core/src/test/resources/example-types.yml
+++ b/conjure-java-core/src/test/resources/example-types.yml
@@ -153,7 +153,7 @@ types:
             docs: docs for optionalItem field
           items:
             type: list<string>
-            docs: docs for items field
+            docs: docs for items field with exciting character$ used by javapoet.
           set:
             type: set<string>
             docs: docs for set field

